File: app/src/main/java/com/example/schoolairdroprefactoredition/scene/user/UserUpdateNameActivity.java
package com.example.schoolairdroprefactoredition.scene.user;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.blankj.utilcode.util.KeyboardUtils;
import com.example.schoolairdroprefactoredition.R;
import com.example.schoolairdroprefactoredition.domain.DomainToken;
import com.example.schoolairdroprefactoredition.domain.DomainBaseUserInfo;
import com.example.schoolairdroprefactoredition.scene.base.ImmersionStatusBarActivity;
import com.example.schoolairdroprefactoredition.utils.ConstantUtil;

import java.lang.reflect.InvocationTargetException;

import javadz.beanutils.BeanUtils;

public class UserUpdateNameActivity extends ImmersionStatusBarActivity implements View.OnClickListener {

    /**
     * @param token 验证信息
     * @param info  我的信息
     */
    public static void start(Context context, DomainToken token, Object info) {
        if (token == null) return;

        DomainBaseUserInfo my = new DomainBaseUserInfo();

        try {
            BeanUtils.copyProperties(my, info);
        } catch (IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        Intent intent = new Intent(context, UserUpdateNameActivity.class);
        intent.putExtra(ConstantUtil.KEY_TOKEN, token);
        intent.putExtra(ConstantUtil.KEY_USER_BASE_INFO, my);
        if (context instanceof AppCompatActivity) {
            ((AppCompatActivity) context).startActivityForResult(intent, UserActivity.REQUEST_UPDATE);
        }
    }

    private ResultViewModel viewModel;

    private EditText mInput;
    private TextView mTitle;
    private TextView mTip;

    private String name;

    private Bundle bundle;

    private DomainToken token;
    private DomainBaseUserInfo info;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_selling_add_set);
        viewModel = new ViewModelProvider(this).get(ResultViewModel.class);

        setSupportActionBar(findViewById(R.id.toolbar));

        mTip = findViewById(R.id.input_tip);
        mTitle = findViewById(R.id.title);
        mInput = findViewById(R.id.input);
        mInput.setInputType(EditorInfo.TYPE_NULL);
        mInput.setLines(1);
        mInput.setOnClickListener(this);
        mInput.requestFocus();
        KeyboardUtils.showSoftInput();

        bundle = getIntent().getExtras();
        if (bundle == null)
            bundle = new Bundle();

        token = (DomainToken) bundle.getSerializable(ConstantUtil.KEY_TOKEN);
        info = (DomainBaseUserInfo) bundle.getSerializable(ConstantUtil.KEY_USER_BASE_INFO);

        init();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = new MenuInflater(this);
        inflater.inflate(R.menu.toolbar_done, menu);
        menu.getItem(0).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            finish();
        } else if (id == R.id.done) {
            rename();
        }
        return super.onOptionsItemSelected(item);
    }

    private void init() {
        mTitle.setText(R.string.setName);

        if (bundle == null)
            bundle = new Bundle();

        token = (DomainToken) bundle.getSerializable(ConstantUtil.KEY_TOKEN);
        info = (DomainBaseUserInfo) bundle.getSerializable(ConstantUtil.KEY_USER_BASE_INFO);
        if (info != null) name = info.getUname();
        mInput.setText(name);
        mInput.setSelection(name.length());
    }

    /**
     * 将数据回到上一个Activity
     */
    private void sendData() {
        info.setUname(mInput.getText().toString().trim());
        bundle.putSerializable(ConstantUtil.KEY_USER_BASE_INFO, info);
        Intent intent = new Intent();
        intent.putExtras(bundle);
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    /**
     * 网络请求修改名字
     */
    private void rename() {
        KeyboardUtils.hideSoftInput(this);

        if (name.equals(mInput.getText().toString().trim()))
            finish();

        if (token != null && token.getAccess_token() != null) {
            viewModel.rename(token.getAccess_token(), mInput.getText().toString().trim())
                    .observe(this, data -> {
                        if (data) {
                            sendData();
                        } else {
                            Toast.makeText(this, R.string.errorUpdatingName, Toast.LENGTH_SHORT).show();
                        }

                    });
        }
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.input) {
            if (!mInput.hasFocus())
                mInput.requestFocus();
        }
    }

}

File: app/src/main/java/com/example/schoolairdroprefactoredition/scene/user/UserUpdateNameActivity.kt
package com.example.schoolairdroprefactoredition.scene.user

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.InputFilter
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.addTextChangedListener
import androidx.lifecycle.ViewModelProvider
import com.blankj.utilcode.util.KeyboardUtils
import com.blankj.utilcode.util.LogUtils
import com.example.schoolairdroprefactoredition.R
import com.example.schoolairdroprefactoredition.application.SAApplication
import com.example.schoolairdroprefactoredition.domain.DomainUserInfo
import com.example.schoolairdroprefactoredition.scene.base.ImmersionStatusBarActivity
import com.example.schoolairdroprefactoredition.utils.ConstantUtil
import com.example.schoolairdroprefactoredition.utils.DialogUtil
import com.example.schoolairdroprefactoredition.utils.MyUtil
import com.example.schoolairdroprefactoredition.utils.RegExrUtil
import com.example.schoolairdroprefactoredition.viewmodel.UserRenameViewModel
import kotlinx.android.synthetic.main.activity_selling_add_set.*
import java.util.regex.Pattern

class UserUpdateNameActivity : ImmersionStatusBarActivity() {

    companion object {

        /**
         * 名字最大长度
         */
        const val MAX_NAME_LENGTH = 8


        fun start(context: Context?) {
            val intent = Intent(context, UserUpdateNameActivity::class.java)
            if (context is AppCompatActivity) {
                context.startActivityForResult(intent, UserActivity.REQUEST_UPDATE_MY)
            }
        }
    }

    private val viewModel by lazy {
        ViewModelProvider(this).get(UserRenameViewModel::class.java)
    }

    /**
     * 完成按钮
     *
     * 当[input_warning]可见时按钮将不可用
     */
    private var doneMenu: MenuItem? = null

    private val loading by lazy {
        MyUtil.loading(this)
    }

    /**
     * 仅匹配中文、字母、数字和下划线
     */
    private val noSpecialsPattern by lazy {
        Pattern.compile(RegExrUtil.NO_SPECIALS)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_selling_add_set)
        setSupportActionBar(findViewById(R.id.toolbar))
        init()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.toolbar_done, menu)
        menu.getItem(0).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        doneMenu = item
        val id = item.itemId
        if (id == android.R.id.home) {
            finish()
        } else if (id == R.id.done) {
            if (input_warning.visibility != View.VISIBLE) {
                rename()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun init() {
        input_attention.visibility = View.GONE
        val myInfo = (application as SAApplication).getCachedMyInfo()
        set_title.setText(R.string.setName)

        input.apply {
            filters = arrayOf(InputFilter.LengthFilter(8))
            input_warning.setText(R.string.illegalName)
            setLines(1)
            addTextChangedListener({ _, _, _, _ -> }, { _, _, _, _ -> }, {
                LogUtils.d("after text changed")
                input_tip.text = getString(R.string.textRemainCount, MAX_NAME_LENGTH - input.text.length)
                if (!noSpecialsPattern.matcher(input.text).matches()) {
                    LogUtils.d("字符输入非法")
                    // 含有特殊字符，禁用提交按钮
                    doneMenu?.isEnabled = false
                    input_warning.visibility = View.VISIBLE
                } else {
                    LogUtils.d("字符输入合法")
                    doneMenu?.isEnabled = true
                    input_warning.visibility = View.INVISIBLE
                }
            })
            setOnFocusChangeListener { v, hasFocus ->
                if (hasFocus) {
                    KeyboardUtils.showSoftInput(v)
                } else {
                    KeyboardUtils.hideSoftInput(v)
                }
            }

            setOnClickListener {
                if (!hasFocus()) {
                    requestFocus()
                }
            }

            setText(myInfo?.userName)
            setSelection(text.length)

            post {
                requestFocus()
            }
        }
    }

    /**
     * 将数据回到上一个Activity
     */
    private fun sendData(myInfo: DomainUserInfo.DataBean?) {
        myInfo?.userName = input.text.toString().trim()
        intent.putExtra(ConstantUtil.KEY_USER_INFO, myInfo)
        setResult(RESULT_OK, intent)
        finish()
    }

    /**
     * 修改用户名
     */
    private fun rename() {
        val myInfo = (application as SAApplication).getCachedMyInfo()
        KeyboardUtils.hideSoftInput(this)
        if (input.text.toString().trim() == myInfo?.userName) {
            finish()
        } else {
            loading.show()
            val token = (application as SAApplication).getCachedToken()
            if (token != null && token.access_token != null) {
                viewModel.rename(token.access_token, input.text.toString().trim())
                        .observeOnce(this, {
                            loading.dismissWith {
                                if (it) {
                                    sendData(myInfo)
                                } else {
                                    DialogUtil.showCenterDialog(this, DialogUtil.DIALOG_TYPE.FAILED, R.string.systemBusy)
                                }
                            }
                        })
            }
        }
    }
}